pymc/aesaraf.py

```diff
@@ -1015,3 +1019,144 @@ def compile_pymc(
+    fg = aesara.graph.FunctionGraph(outputs=outputs, clone=False)
+
+    # Walk the graph from inputs to outputs and tag the volatile variables
+    nodes = general_toposort(fg.outputs, deps=lambda x: x.get_parents())
+    givens = []
+    registered_givens = set()
+    volatile_nodes = set()
+    for node in nodes:
+        if (
+            isinstance(node, aesara.compile.SharedVariable)
+            or node in fg.outputs
+            or node in givens_dict
+        ):
+            volatile_nodes.add(node)
+        is_volatile = node in volatile_nodes
+
+        # Populate the givens list if the node is in the givens_dict
+        if node in givens_dict and node not in registered_givens:
+            value = givens_dict[node]
+            if not isinstance(value, (Apply, Variable)):
+                # If the value is not a tensor, it is wrapped as a Constant with the node's name
+                value = at.constant(
+                    givens_dict[node],
+                    dtype=node.dtype,
+                    name=node.name,
+                )
+            givens.append((node, value))
+            registered_givens.add(node)
+
+        # Propagate the volatile property to the node's clients
+        if is_volatile:
+            try:
+                clients = [c for c, _ in fg.clients[node]]
+            except KeyError:
+                # Some Apply nodes don't show up in the clients of a FunctionGraph
+                # so we take the Apply node's outputs instead
+                if isinstance(node, aesara.graph.basic.Apply):
+                    clients = node.outputs
+                else:
+                    raise
+            for client in clients:
+                if (
+                    isinstance(client, aesara.graph.basic.Variable)
+                    and client.owner is not None
+                    and isinstance(client.owner.op, at.random.basic.RandomVariable)
+                    and any(rv_param in volatile_nodes for rv_param in client.owner.inputs[3:])
+                ):
+                    # Random variables that have volatile parameters are also volatile
+                    volatile_nodes.add(client)
+                elif not (
+                    client in vars_in_trace
+                    and isinstance(client, aesara.graph.basic.Variable)
+                    and client.owner is not None
+                    and isinstance(client.owner.op, at.random.basic.RandomVariable)
+                ):
+                    # Other variables that aren't random variables in the trace are volatile
+                    volatile_nodes.add(client)
+
+    # Collect the function inputs by walking the graph from the outputs. Inputs will be:
+    # 1. Random variables that are not volatile
+    # 2. Variables that have no owner and are not constant or shared
+    inputs = []
+
+    def expand(node):
+        if isinstance(node, Apply):
+            # Apply nodes are never inputs
+            return node.get_parents()
```

Comment:
Why are we including them in the expansion at all?

Reply:
What do you mean?

Reply:
I don't get why we are getting `Apply` nodes in the expansion function. Is this something that `node.get_parents()` does? What about expanding with `return node.owner.inputs` instead?

Reply:
I though that `get_parents` just gave the `owner` if the node was a `Variable` or the `inputs` if the node was an `Apply`

Reply:
Yes I think that's what it does, but why do we need the owner of the Variable? That will always be the end of the graph in that branch

Reply:
I don't think I followed you, don't we need to expand the full graph through the Variable owners?

Reply:
I added a change suggestion below that explains why we don't need to care about applys in the graph expansion